### PR TITLE
[fix] Support Timesformer feature extraction

### PR DESCRIPTION
**Pull request by SCZwangxiao**
## Motivation

Currently, `Timesformer` is not compatible with the feature extraction module in `recognizer3D`.

The current feature extraction module in `recognizer3D` is illustrated below:
```python
        if self.feature_extraction:
            # perform spatio-temporal pooling
            avg_pool = nn.AdaptiveAvgPool3d(1)
            if isinstance(feat, tuple):
                feat = [avg_pool(x) for x in feat]
                # concat them
                feat = torch.cat(feat, axis=1)
            else:
                feat = avg_pool(feat)
            # squeeze dimensions
            feat = feat.reshape((batches, num_segs, -1))
            # temporal average pooling
            feat = feat.mean(axis=1)
            return feat
```
This module is only compatible with `feat` of shape `(N, in_channels, T, H, W)`. However,`feat` in Timesformer are of shape `(N, in_channels)`, which requires no Pooling.

## Modification
Add different branches according to `feat` shape.

### CODE REVIEW

**Comment by dreamerlin** [comment]
Need to use pre-commit to fix the lint problem, ref: https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md

**Comment by dreamerlin** [comment]
pre-commit run --all-files

File: mmaction/models/recognizers/recognizer3d.py [hunk 1]
```diff
@@ -64,18 +64,24 @@ def _do_test(self, imgs):
...
+                'only 3D-CNN-like ([N, in_channels, T, H, W]), and '
+                'transformer-like ([N, in_channels]) features are supported.')
+            if len(feat.size()) == 5:  # 3D-CNN architecture
+                # perform spatio-temporal pooling
+                avg_pool = nn.AdaptiveAvgPool3d(1)
+                if isinstance(feat, tuple):
+                    feat = [avg_pool(x) for x in feat]
+                    # concat them
+                    feat = torch.cat(feat, axis=1)
+                else:
```
Comment (congee524) [inline mmaction/models/recognizers/recognizer3d.py:79]: `elif len(feat.size()) == 2`

**Review by congee524 — APPROVED** [review]

File: mmaction/models/recognizers/recognizer3d.py [hunk 2]
```diff
@@ -64,18 +64,24 @@ def _do_test(self, imgs):
...
-                feat = [avg_pool(x) for x in feat]
-                # concat them
-                feat = torch.cat(feat, axis=1)
-            else:
-                feat = avg_pool(feat)
-            # squeeze dimensions
-            feat = feat.reshape((batches, num_segs, -1))
-            # temporal average pooling
-            feat = feat.mean(axis=1)
+            assert len(feat.size()) in [
```
Comment (dreamerlin) [inline mmaction/models/recognizers/recognizer3d.py:67]: `feat` may be a tuple in TPN

File: mmaction/models/recognizers/recognizer3d.py [hunk 3]
```diff
@@ -64,18 +64,24 @@ def _do_test(self, imgs):
...
+                5, 2
+            ], ('Got feature of unknown architecture, '
+                'only 3D-CNN-like ([N, in_channels, T, H, W]), and '
+                'transformer-like ([N, in_channels]) features are supported.')
+            if len(feat.size()) == 5:  # 3D-CNN architecture
```
Comment (dreamerlin) [inline mmaction/models/recognizers/recognizer3d.py:72]: `feat` may be a tuple in TPN

**Review by dreamerlin — APPROVED** [review]

**Comment by kennymckormick** [comment]
Tested.

**Pull request merged**